### PR TITLE
Fix public links actions

### LINKED ISSUE #1891: file operations are not possible with public links
### Steps to reproduce
1. create a public link of a folder with "Editor" or "Contributor" role
2. try to upload/rename/delete files through the public link

### Expected behaviour
file operations can be done when the permissions are given

### Actual behaviour
not file-openrations can be done, buttons are greyed out
image

### PR DESCRIPTION
**Pull request by LukasHirt**
## Description
Disable quota for public links and pass header for upload in password-protected links.

## Related Issue

- Fixes #1891

## How Has This Been Tested?

- test environment: Manually
1. Rename file in folder shared with public link
2. Delete file in folder shared with public link
3. Upload file into folder shared with public link

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by PVince81** [comment]
Doesn't work fully:

1. Create folder
1. Create link on folder with role "Editor" and set a password
1. Open link
1. Enter password
1. Create folder :white_check_mark: 
1. Upload file :white_check_mark: 
1. Download file :white_check_mark: 
1. Rename folder/file :no_entry_sign: it's grayed out
1. Delete folder/file :no_entry_sign: it's grayed out

**Comment by individual-it** [comment]
deleting a file with password protected link does not work
```
error: /yLLig11l5lUQLZs/block-aligned3.txt not deleted: Error: No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured, No public access to this resource., No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured, No 'Authorization: Bearer' header found. Either the client didn't send one, or the server is mis-configured, No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured

**Comment by individual-it** [comment]
@PVince81 try as "Collaborator" I think there is a permission/description mismatch see https://github.com/owncloud/phoenix/issues/1979

**Comment by LukasHirt** [comment]
> deleting a file with password protected link does not work

Seems we are sending password as a wrong argument. https://github.com/owncloud/owncloud-sdk/blob/36e131b085b1a2f4ecb6449a88dc5a471ebd4bcf/src/publicLinkAccess.js#L142

https://github.com/owncloud/phoenix/blob/b936e6378830b291afb704f9b05f91c4f30afd81/apps/files/src/store/actions.js#L448

**Comment by PVince81** [comment]
> @PVince81 try as "Collaborator" I think there is a permission/description mismatch see #1979

Yes, correct. I've adjusted my test results now.

**Comment by PVince81** [comment]
I've rebased and it seems it's more broken than before, possibly because of those JS issues.
Test report here https://github.com/owncloud/phoenix/pull/1951#issuecomment-531702628

**Comment by PVince81** [comment]
seems `currentFolder` is null which is likely the reason the perms are not read correctly.

I'll investigate...

**Comment by PVince81** [comment]
Fixed currentFolder issue and download issue.

Test report updated.

Next up: the delete issue with password

**Comment by PVince81** [comment]
Everything fixed.

Now ideally we should also have the acceptance tests in this PR.

@LukasHirt @DeepDiver1975 please review the code changes

**Comment by PVince81** [comment]
I'm looking into acceptance tests and found another issue with deletion of folders

File: apps/files/src/components/FilesAppBar.vue [hunk 1]
```diff
@@ -5,9 +5,9 @@
       <div class="uk-width-expand">
         <div class="uk-flex">
           <oc-breadcrumb id="files-breadcrumb" :items="breadcrumbs" v-if="showBreadcrumb" home></oc-breadcrumb>
-          <span class="uk-margin-small-left" v-if="showBreadcrumb && currentFolder.privateLink">
+          <span class="uk-margin-small-left" v-if="showBreadcrumb && currentFolder && currentFolder.privateLink">
           <oc-icon name="ready" v-show="linkCopied" />
-          <oc-icon id="files-permalink-copy" name="link" v-clipboard="() => currentFolder.privateLink"
+          <oc-icon id="files-permalink-copy" name="link" v-clipboard="() => currentFolder && currentFolder.privateLink"
```
Comment (LukasHirt) [inline apps/files/src/components/FilesAppBar.vue]: Is the check for current folder here necessary when we check for it already in the parent?

Comment (PVince81) [reply apps/files/src/components/FilesAppBar.vue]: yes, this solves the JS error. I guess the vue was still rendering or evaluating this piece while the current folder was not loaded yet

Comment (PVince81) [reply apps/files/src/components/FilesAppBar.vue]: ah, got your point. I didn't check

**Comment by PVince81** [comment]
I wanted to try adding tests but realized that I need a logout step, so I raised another PR for that first: https://github.com/owncloud/phoenix/pull/2100

**Comment by PVince81** [comment]
as discussed, no need for logout step as we can create the link using the API.

I tried just that locally and now bumped into https://github.com/owncloud/phoenix/issues/1882 when running the tests

**Comment by PVince81** [comment]
I've added one test for public link. This should be the base work for adding more.

**Comment by PVince81** [comment]
@individual-it please review the tests

File: apps/files/src/components/FilesAppBar.vue (hunk 1 above)
Comment (PVince81) [reply apps/files/src/components/FilesAppBar.vue]: fixed

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 2]
```diff
@@ -332,6 +372,10 @@ module.exports = {
       selector: '//button[@aria-label="show-file-actions"]',
       locateStrategy: 'xpath'
     },
+    fileActionsButtonDisabledInFileRow: {
```
Comment (individual-it) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: unused

File: tests/acceptance/pageObjects/filesPage.js [hunk 3]
```diff
@@ -74,6 +74,20 @@ module.exports = {
...
+    /**
+     * Returns whether files or folders can be created in the current page.
+     *
+     * @param {Function} callback callback with result
+     */
+    canCreateFiles: function (callback) {
+      return this
+        .waitForElementVisible('@newFileMenuButtonAnyState')
+        .getAttribute('@newFileMenuButtonAnyState', 'disabled', (result) => {
+          console.log(result)
```
Comment (individual-it) [inline tests/acceptance/pageObjects/filesPage.js]: debug output?

File: tests/acceptance/pageObjects/FilesPageElement/filesList.js [hunk 4]
```diff
@@ -36,6 +36,46 @@ module.exports = {
...
+     * @param {function(boolean)} callback - whether the action is disabled
+     * @returns {*}
+     */
+    isActionDisabled: function (fileName, action, callback) {
+      const btnSelectorHighResolution = '(' + this.getFileRowSelectorByFileName(fileName) +
+        this.elements[action + 'ButtonInFileRow'].selector + ')[1]'
+      const btnSelectorLowResolution = '(' + this.getFileRowSelectorByFileName(fileName) +
+        this.elements[action + 'ButtonInFileRow'].selector + ')[last()]'
+      const fileActionsBtnSelector = this.getFileRowSelectorByFileName(fileName) +
+        this.elements.fileActionsButtonInFileRow.selector
```
Comment (individual-it) [inline tests/acceptance/pageObjects/FilesPageElement/filesList.js]: put in separate function to generate this selectors, so it also used for https://github.com/owncloud/phoenix/pull/1951/commits/195d9b18222db8618d6ee5176a7989346f07e522#diff-2da1b8066c5b1ccfa1ef0555b24164dbL16

Comment (PVince81) [reply tests/acceptance/pageObjects/FilesPageElement/filesList.js]: done

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUISharingPublic</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/5535/

<p>





</p></details>

**Comment by PVince81** [comment]
fixed failing test, it was due to a collision of assumptions when rebasing this PR

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>XGAPortrait</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/5543/

<p>




</p></details>

**Comment by PVince81** [comment]
@individual-it please review again. Can we ignore the XGA failure ?

File: tests/acceptance/stepDefinitions/filesContext.js [hunk 5]
```diff
@@ -141,14 +141,32 @@ Given('the following files have been deleted by user {string}', async function (
...
+
+Then('it should not be possible to delete file {string} using the webUI', function (resource) {
+  return client.page.FilesPageElement.filesList().isActionDisabled(resource, 'delete', (isDisabled) => {
+    client.assert.strictEqual(isDisabled, true, `Delete action must not be enabled for "${resource}"`)
+  })
+})
 
+Given('the following files have been deleted by user {string}', function (user, table) {
+  for (const line of table.rows()) {
+    webdav.delete(user, line[0])
```
Comment (individual-it) [inline tests/acceptance/stepDefinitions/filesContext.js]: Use await to make sure files have been finished deleting before the next step starts

**Comment by PVince81** [comment]
ah, probably a conflict resolution fail

**Comment by PVince81** [comment]
solved duplicate step definition, it was a conflict resolution fail.

File: tests/acceptance/stepDefinitions/filesContext.js (hunk 5 above)
Comment (PVince81) [reply tests/acceptance/stepDefinitions/filesContext.js]: obsolete

**Comment by PVince81** [comment]
Conflict resolved

File: tests/acceptance/features/webUISharingPublic/shareByPublicLink.feature [hunk 6]
```diff
@@ -84,13 +85,33 @@ Feature: Share by public link
...
+  Scenario: creating a public link with "Viewer" role only makes it impossible to delete files via the link
+    Given user "user1" has shared folder "simple-folder" with link with "read" permissions
+    When the public uses the webUI to access the last public link created by user "user1"
     Then it should not be possible to delete file "lorem.txt" using the webUI
 
+  Scenario: creating a public link with "Editor" role makes it possible to upload files via the link even with password set
+    Given user "user1" has shared folder "simple-folder" with link with "read, change, create, delete" permissions and password "pass123"
+    When the public uses the webUI to access the last public link created by user "user1" with password "pass123"
+    And the user uploads file "new-lorem.txt" using the webUI
+    Then file "new-lorem.txt" should be listed on the webUI
```
Comment (individual-it) [inline tests/acceptance/features/webUISharingPublic/shareByPublicLink.feature]: you could add 
```  
    And as "user1" file "new-lorem.txt" should exist
```

just to make sure the file really is available for the user and not only the UI lying

File: tests/acceptance/pageObjects/publicLinkPasswordPage.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      submitPublicLinkPassword: async function (password) {
+        await this
+          .waitForElementVisible('@passwordInput')
+          .setValue('@passwordInput', password)
+          .click('@passwordSubmitButton')
+
+        return this
+          .page.FilesPageElement.filesList()
+          .waitForElementPresent({ selector: '@filesListProgressBar', abortOnFailure: false }) // don't fail if we are too late
+          .waitForElementNotPresent('@filesListProgressBar')
```
Comment (individual-it) [inline tests/acceptance/pageObjects/publicLinkPasswordPage.js]: with #2082 merged you can now use
```
      return navigationHelper.navigateAndWaitTillLoaded(
        this.url(), this.page.FilesPageElement.filesList().elements.filesListProgressBar.selector
      )
```
see https://github.com/owncloud/phoenix/pull/2082/commits/9f7cac7241f57be5091d63d0e90b999975fa0708

File: tests/acceptance/features/webUISharingPublic/shareByPublicLink.feature (hunk 6 above)
Comment (PVince81) [reply tests/acceptance/features/webUISharingPublic/shareByPublicLink.feature]: uh oh
```
   ✔ Then file "new-lorem.txt" should be listed on the webUI # tests/acceptance/stepDefinitions/filesContext.js:214
   ✖ And as "user1" file "new-lorem.txt" should exist # tests/acceptance/stepDefinitions/webdavContext.js:8
       Error: file/folder should exist, but does not
```

I'll debug...

Comment (PVince81) [reply tests/acceptance/features/webUISharingPublic/shareByPublicLink.feature]: Fixed. the path must be "simple-folder/new-lorem.txt" in that check

**Comment by PVince81** [comment]
tests passed but unpublished. => restarted build

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>webUIFiles</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/5678/

<p>





</p></details>

**Comment by PVince81** [comment]
Ran the failing tests locally and they work fine.

Attempting another rebuild now...

**Comment by PVince81** [comment]
Resolved once again an ambiguous test which came from master through rebasing.

Hoping for a full pass now and no more conflicts...

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Trashbin</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/5705/

<p>



















</p></details>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Files</strong> failed. Please find the screenshots inside ...</summary>

https://drone.owncloud.com/owncloud/phoenix/5705/

<p>










































</p></details>

**Comment by PVince81** [comment]
more failures :'(

**Comment by PVince81** [comment]
both pass locally... trying my luck again...

I hope it's not anything I changed that is now triggering these random failures ?

**Pull request merged**